internal/object/object.go

```diff
@@ -24,28 +25,84 @@ var (
 )
 
 type Object struct {
-	ID       types.Binary
-	SourceId int64
-	Name     string
-	Tags     map[string]string
-	URL      string
+	ID       types.Binary `db:"id"`
+	SourceID int64        `db:"source_id"`
+	Name     string       `db:"name"`
+	URL      types.String `db:"url"`
 
+	Tags      map[string]string
 	ExtraTags map[string]string
 
 	db *icingadb.DB
 }
 
-func NewObject(db *icingadb.DB, ev *event.Event) *Object {
+// New creates a new object from the given event.
+func New(db *icingadb.DB, ev *event.Event) *Object {
 	return &Object{
-		SourceId:  ev.SourceId,
+		SourceID:  ev.SourceId,
 		Name:      ev.Name,
 		db:        db,
-		URL:       ev.URL,
+		URL:       utils.ToDBString(ev.URL),
 		Tags:      ev.Tags,
 		ExtraTags: ev.ExtraTags,
 	}
 }
 
+// GetFromCache fetches an object from the global object cache store matching the given ID.
+// Returns nil if it's not in the cache.
+func GetFromCache(id types.Binary) *Object {
+	cacheMu.Lock()
+	defer cacheMu.Unlock()
+
+	return cache[id.String()]
+}
+
+// RestoreObjects restores all objects and their (extra)tags matching the given IDs from the database.
+// Returns error on any database failures and panics when trying to cache an object that's already in the cache store.
+func RestoreObjects(ctx context.Context, db *icingadb.DB, ids []any) error {
+	err := utils.ForEachRow[Object](ctx, db, "id", ids, func(o *Object) {
+		o.db = db
+		o.Tags = map[string]string{}
+		o.ExtraTags = map[string]string{}
+
+		cacheMu.Lock()
+		defer cacheMu.Unlock()
+
+		if obj, ok := cache[o.ID.String()]; ok {
+			panic(fmt.Sprintf("Object %q is already in the cache", obj.DisplayName()))
+		}
+
+		cache[o.ID.String()] = o
+	})
+	if err != nil {
+		return errors.Wrap(err, "cannot restore incidents object")
+	}
+
+	// Restore object ID tags matching the given object ids
+	err = utils.ForEachRow[IdTagRow](ctx, db, "object_id", ids, func(ir *IdTagRow) {
+		cacheMu.Lock()
+		defer cacheMu.Unlock()
+
+		cache[ir.ObjectId.String()].Tags[ir.Tag] = ir.Value
+	})
+	if err != nil {
+		return errors.Wrap(err, "cannot restore incident object ID tags")
+	}
+
+	// Restore object extra tags matching the given object ids
+	err = utils.ForEachRow[ExtraTagRow](ctx, db, "object_id", ids, func(et *ExtraTagRow) {
+		cacheMu.Lock()
+		defer cacheMu.Unlock()
+
+		cache[et.ObjectId.String()].ExtraTags[et.Tag] = et.Value
+	})
+	if err != nil {
+		return errors.Wrap(err, "cannot restore incident object extra tags")
+	}
+
+	return nil
+}
```

Comment:
This locks `cacheMu` quite a few times. It would make sense to create a local map first, and move all the entries to the shared `cache` at the end. Not only to avoid extra locking but also that there are no half-initialized objects in `cache` (which may even be left there if the function would return with an error.